pkg/interceptors/server/server.go

```diff
@@ -0,0 +1,123 @@
+package server
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/tektoncd/triggers/pkg/interceptors/bitbucket"
+	"github.com/tektoncd/triggers/pkg/interceptors/cel"
+	"github.com/tektoncd/triggers/pkg/interceptors/github"
+	"github.com/tektoncd/triggers/pkg/interceptors/gitlab"
+
+	"github.com/tektoncd/triggers/pkg/apis/triggers/v1alpha1"
+	"go.uber.org/zap"
+	"k8s.io/client-go/kubernetes"
+)
+
+type Server struct {
+	KubeClient   kubernetes.Interface
+	Logger       *zap.SugaredLogger
+	interceptors map[string]v1alpha1.InterceptorInterface
+}
+
+func NewWithCoreInterceptors(k kubernetes.Interface, l *zap.SugaredLogger) *Server {
+	i := map[string]v1alpha1.InterceptorInterface{
+		"bitbucket": bitbucket.NewInterceptor(k, l).(v1alpha1.InterceptorInterface),
+		"cel":       cel.NewInterceptor(k, l).(v1alpha1.InterceptorInterface),
+		"github":    github.NewInterceptor(k, l).(v1alpha1.InterceptorInterface),
+		"gitlab":    gitlab.NewInterceptor(k, l).(v1alpha1.InterceptorInterface),
+	}
+	s := Server{
+		KubeClient:   k,
+		Logger:       l,
+		interceptors: i,
+	}
+	return &s
+}
+
+func (is *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	b, err := is.ExecuteInterceptor(r)
+	if err != nil {
+		switch e := err.(type) {
+		case Error:
+			is.Logger.Infof("HTTP %d - %s", e.Status(), e)
+			http.Error(w, e.Error(), e.Status())
+		default:
+			is.Logger.Errorf("Non Status Error: %s", err)
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		}
+	}
+	w.Header().Add("Content-Type", "application/json")
+	if _, err = w.Write(b); err != nil {
+		is.Logger.Errorf("failed to write response: %s", err)
```

Comment:
I think we need to explicitly set the response code - I'm not sure if this is set automatically if the write fails.

Reply:
The first call to Write() will automatically call WriteHeader(200) if WriteHeader has not been called before: 

```
	// If WriteHeader is not called explicitly, the first call to Write
	// will trigger an implicit WriteHeader(http.StatusOK).
	// Thus explicit calls to WriteHeader are mainly used to
	// send error codes.
```

https://pkg.go.dev/net/http#ResponseWriter

Reply:
Yup! I don't know if the same holds true if the call to w.Write fails though - we should probably return a 500 or similar. I forget what the behavior is if we reach the end of the handler without and successful writes - I want to make sure we don't accidentally return OK if the write fails.

Reply:
oh sorry, I missed this comment yesterday -- my understanding is that the headers and status code gets flushed first before the body is written. So, if we the body write fails somewhere in the middle, we don't really have a way to send 500s back anymore (since the status code has already been sent). The best we can do is log the error on the server side.